src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php

```diff
@@ -125,7 +128,7 @@ public function preSubmit(FormEvent $event)
                 if (!$form->has($name)) {
                     $form->add($name, $this->type, array_replace(array(
                         'property_path' => '['.$name.']',
-                    ), $this->options));
+                    ), $this->getOptionsForEntry($value)));
```

Comment:
It's funny because I was thinking about it last night and I wondered how was the `prototype_data` option handled though, and I realised while remembering your comment that you can only define the view data of a prototype. The `CollectionType` doc reference should be clear about it, especially if you need to transform value.

And just this morning, I see this ticket #17992...

Since there is no problem as long as the underlying data is a string value, it may not be working for an object underlying data.

I see an easy solution but it's not "pretty", it should be mentioned in the docs that the `entry_options` option as callable must test the string type :

```php
'entry_options' => function ($entry) {
    $options = array(
        // set default
    );

    if (is_string($entry) {
        // $entry <=> $viewData
        // or $entry <=> $modelData <=> $viewData
        $options['global_dynamic'] = // use $entry as string
    }

    // Else entry is an instanceof `data_class` or null
    $options['model_dynamic'] = // use $entry as model data or null

    return $options;
},
```

But it would be better to solve this in a feature way, by adding for example an `add_entry` option (which may have its counterpart in a future `allow_add`'able `ChoiceType` as `add_choice`).

This option would be a callable getting passed the new view data as only argument and should return a new model data. Almost like `empty_data` but this last one is only called for an empty view data and takes the form as first argument.

Then we would be able to use `add_entry` call on the new view data before passing it to `getOptionsForEntry`.

Another solution is to try to fix this while refactoring `CollectionType` more globally thanks to a `CollectionList` support, since I'm already investigating this way.
I think this would allow to solve many old issues with this type, with at first dynamic indexing, and polymorphism.

What do you think ?